concordium-std/src/impls.rs

```diff
@@ -575,61 +576,76 @@ fn prepend_generic_map_key<K: Serial>(key: K) -> Vec<u8> {
 
-    fn entry(&mut self, key: &[u8]) -> EntryRaw<Self::EntryType> {
-        let key_start = key.as_ptr();
-        let key_len = key.len() as u32; // Wasm usize == 32bit.
-        let res = unsafe { lookup(key_start, key_len) };
+    /// Open the contract state.
+    fn open(data: T) -> Self {
+        Self::new(data)
+    }
 
-        if res == -1 {
-            let entry_id = unsafe { create(key_start, key_len) };
-            EntryRaw::Vacant(VacantEntryRaw::new(StateEntry::open(entry_id)))
-        } else {
-            let entry_id = res as u32;
-            EntryRaw::Occupied(OccupiedEntryRaw::new(StateEntry::open(entry_id)))
+    fn entry(&mut self, key: &[u8]) -> EntryRaw<<T as ContractStateApi>::EntryType> {
+        match self.state_api.lookup(key) {
+            Some(entry) => EntryRaw::Occupied(OccupiedEntryRaw::new(entry)),
+            None => {
+                let entry = self.state_api.create(key);
+                EntryRaw::Vacant(VacantEntryRaw::new(entry))
+            }
         }
     }
 
-    fn lookup(&self, key: &[u8]) -> Option<Self::EntryType> {
-        let key_start = key.as_ptr();
-        let key_len = key.len() as u32; // Wasm usize == 32bit.
-        let entry_id_option = unsafe { lookup(key_start, key_len) };
-        if entry_id_option < 0 {
-            None
-        } else {
-            Some(StateEntry::open(entry_id_option as u32))
-        }
+    fn lookup(&self, key: &[u8]) -> Option<<T as ContractStateApi>::EntryType> {
+        self.state_api.lookup(key)
     }
 
     /// Delete the entry. Returns true if the entry was occupied and false
     /// otherwise.
-    fn delete_entry(&mut self, entry: Self::EntryType) -> bool {
-        unsafe { delete_entry(entry.state_entry_id) == 1 }
+    fn delete_entry(&mut self, entry: <T as ContractStateApi>::EntryType) -> bool {
+        self.state_api.delete_entry(entry)
     }
 
     /// If exact, delete the specific key, otherwise delete the subtree.
     /// Returns true if entry/subtree was occupied and false otherwise
     /// (including if the key was too long or empty).
-    fn delete_prefix(&mut self, prefix: &[u8], exact: bool) -> bool {
-        let len = prefix.len() as u32; // Safe because usize is 32bit in WASM.
-        let prefix_ptr = prefix.as_ptr();
-        unsafe { delete_prefix(prefix_ptr, len, exact as u32) == 1 }
+    fn delete_prefix(&mut self, prefix: &[u8]) -> bool {
+        self.state_api.delete_prefix(prefix)
     }
 
-    fn iterator(&self, prefix: &[u8]) -> Self::IterType {
-        let prefix_start = prefix.as_ptr();
-        let prefix_len = prefix.len() as u32; // Wasm usize == 32bit.
-        let iterator_id = unsafe { iterator(prefix_start, prefix_len) };
-        ContractStateIter {
-            iterator_id,
+    fn iterator(
+        &self,
+        prefix: &[u8],
+    ) -> Result<<T as ContractStateApi>::IterType, ContractStateError> {
+        let locked = self.locked_subtrees.borrow().keys().any(|p| {
```

Comment:
The contract should not be checking whether a subtree is locked on its own. 
It should just call `iterator()` on the host and parse the response.

Reply:
But the test trie  _should_ perform these kinds of checks so that the semantics are identical to those of the host.